tpke/src/ciphertext.rs

```diff
@@ -82,31 +90,57 @@ pub fn decrypt<E: PairingEngine>(
     privkey: E::G2Affine,
 ) -> Vec<u8> {
     let s = E::product_of_pairings(&[(
-        E::G1Prepared::from(ciphertext.nonce),
+        E::G1Prepared::from(ciphertext.commitment),
         E::G2Prepared::from(privkey),
     )]);
     decrypt_with_shared_secret(ciphertext, &s)
 }
-pub fn decrypt_with_shared_secret<E: PairingEngine>(
+
+fn decrypt_with_shared_secret<E: PairingEngine>(
     ciphertext: &Ciphertext<E>,
     s: &E::Fqk,
 ) -> Vec<u8> {
-    let mut plaintext = ciphertext.ciphertext.to_vec();
-    let mut cipher = shared_secret_to_chacha::<E>(s);
-    cipher.apply_keystream(&mut plaintext);
+    let nonce = commitment_to_nonce::<E>(ciphertext.commitment);
+    let ciphertext = ciphertext.ciphertext.to_vec();
+
+    let cipher = shared_secret_to_chacha::<E>(s);
+    let plaintext = cipher.decrypt(&nonce, ciphertext.as_ref()).unwrap();
 
     plaintext
 }
 
-pub fn shared_secret_to_chacha<E: PairingEngine>(s: &E::Fqk) -> ChaCha20 {
+pub fn checked_decrypt_with_shared_secret<E: PairingEngine>(
```

Comment:
could we add a test for this new function? I know that there are some end to end tests in `lib.rs`, but I think we should start expanding the test suite, especially with anything we add